server/src/main/java/com/genymobile/scrcpy/CodecOptions.java

```diff
@@ -0,0 +1,85 @@
+package com.genymobile.scrcpy;
+
+import android.media.MediaCodecInfo;
+import android.os.Build;
+
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+public class CodecOptions {
```

Comment:
(so this class would not be needed at all)

Reply:
I don't think its a smart idea to let the user pass down any key=value pair he wants...
He can totally mess up and give random values and things that might even break the encoding and he won't be aware that he was giving something wrong.

I don't know about the other options but in case of profile and level it is suggested that if you provide a profile, you will also set a level, this is why when only profile (without level) is provided by the user, there is a calculation based on the screen size that should decide the level.

My goal was not to allow the user to pass down any key=value pair he wants, but to allow you or whoever wants to, to expand on the current CodecOptions and easily add new options that he can choose how they will be parsed and what will happen with them and where in the code.

Reply:
> I don't think its a smart idea to let the user pass down any key=value pair he wants...

It would allow to configure the codec generically, without having to build the app.


> He can totally mess up and give random values and things that might even break the encoding and he won't be aware that he was giving something wrong.

At least, we can warn if the provided key is not supported: [`containsKey()`](https://developer.android.com/reference/android/media/MediaFormat#containsKey(java.lang.String)).

User can already pass random values (a bitrate of 1Mbps for instance), which is "ignored" by the codec without warnings.



> I don't know about the other options but in case of profile and level it is suggested that if you provide a profile, you will also set a level, this is why when only profile (without level) is provided by the user, there is a calculation based on the screen size that should decide the level.

Is it done automatically if you don't provide the level? https://github.com/Genymobile/scrcpy/pull/1226#issuecomment-601596309

Btw, contrary to what I said [here](https://github.com/Genymobile/scrcpy/pull/1226#issuecomment-601612659), there is no need to specify the type in the string, it can be requested: [`getValueTypeForKey`](https://developer.android.com/reference/android/media/MediaFormat#getValueTypeForKey(java.lang.String))

Reply:
OK, it will be done your way :)
Thanks for the good lesson.

Reply:
@rom1v 
Just a quick question, what if the key is supported but value is not supported by the device?
If we are assuming the codec ignores bad values shouldn't the user know about it somehow?

Reply:
getValueTypeForKey is API29+ ...
What do you suggest?
We can go back to what you said, assuming int unless specified key:type=value

Reply:
@rom1v Ping

Reply:
> what if the key is supported but value is not supported by the device?

The device will just ignore it I guess. Like if for now you pass `--max-fps` but the encoder does not support it.

> getValueTypeForKey is API29+

:(

> We can go back to what you said, assuming int unless specified key:type=value

Hmm, yes... I have no better solution then :(

Reply:
OK, so I've made the changes on our side, ill need to rebase my dev branch and make a new PR.
@rom1v 
Just to make things clear before i make a new PR, the expected value types are
Integer/int - String/str - Float/float - Long/long
Yes?